ccl/changefeedccl: TestChangefeedSchemaChangeBackfillCheckpoint failed

ccl/changefeedccl.TestChangefeedSchemaChangeBackfillCheckpoint [failed](https://tanzanite.cluster.engflow.com/invocations/default/19c8ffd0-9d77-4d6c-8841-c46b6cfec670?testReportRun=2&testReportShard=10&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [3aa5bdf40527e2f60b179094403b4302a1c2cbe1](https://github.com/cockroachdb/cockroach/commits/3aa5bdf40527e2f60b179094403b4302a1c2cbe1):


```
=== RUN   TestChangefeedSchemaChangeBackfillCheckpoint
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestChangefeedSchemaChangeBackfillCheckpoint295383793
    test_log_scope.go:81: use -show-logs to present logs inline
    changefeed_test.go:2184: random seed: 7283359818651850375
    changefeed_test.go:2450: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedSchemaChangeBackfillCheckpoint295383793
--- FAIL: TestChangefeedSchemaChangeBackfillCheckpoint (261.39s)
=== RUN   TestChangefeedSchemaChangeBackfillCheckpoint/enterprise
    helpers_test.go:867: making server as system tenant
    helpers_test.go:943: making enterprise feed factory
    helpers_test.go:957: pgURL enterprise root
    changefeed_test.go:2421: condition failed to evaluate within 3m45s: from changefeed_test.go:2426: checkpoint still non-empty
    --- FAIL: TestChangefeedSchemaChangeBackfillCheckpoint/enterprise (261.10s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=10</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #119375 ccl/changefeedccl: TestChangefeedSchemaChangeBackfillCheckpoint failed [A-cdc C-test-failure O-robot P-2 T-cdc branch-release-23.1]
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedSchemaChangeBackfillCheckpoint.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39581

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -2205,13 +2205,13 @@ func TestChangefeedSchemaChangeBackfillCheckpoint(t *testing.T) {
 			Changefeed.(*TestingKnobs)
 
 		// Initialize table
-		sqlDB.Exec(t, `CREATE TABLE foo(key INT PRIMARY KEY DEFAULT unique_rowid(), val INT)`)
-		sqlDB.Exec(t, `INSERT INTO foo (val) SELECT * FROM generate_series(1, 1000)`)
+		sqlDB.Exec(t, `CREATE TABLE foo(key INT PRIMARY KEY)`)
+		sqlDB.Exec(t, `INSERT INTO foo (key) SELECT * FROM generate_series(1, 1000)`)
 
 		// Ensure Scan Requests are always small enough that we receive multiple
 		// resolved events during a backfill
 		knobs.FeedKnobs.BeforeScanRequest = func(b *kv.Batch) error {
-			b.Header.MaxSpanRequestKeys = 10
+			b.Header.MaxSpanRequestKeys = 50
 			return nil
 		}
 
@@ -2291,6 +2291,7 @@ func TestChangefeedSchemaChangeBackfillCheckpoint(t *testing.T) {
 				// change due to a primary index swap.
 				refreshTableSpan()
 				backfillTimestamp = r.Timestamp
+				t.Logf("backfill timestamp: %s", backfillTimestamp)
 				return false, nil
 			}
 
@@ -2309,7 +2310,7 @@ func TestChangefeedSchemaChangeBackfillCheckpoint(t *testing.T) {
 			// Filter non-backfill-related spans
 			if !r.Timestamp.Equal(backfillTimestamp.Next()) {
 				// Only allow spans prior to a valid backfillTimestamp to avoid moving past the backfill
-				return !(backfillTimestamp.IsEmpty() || r.Timestamp.LessEq(backfillTimestamp.Next())), nil
+				return backfillTimestamp.IsSet() && !r.Timestamp.LessEq(backfillTimestamp.Next()), nil
 			}
 
 			// At the end of a backfill, kv feed will emit a resolved span for the whole table.
@@ -2337,6 +2338,7 @@ func TestChangefeedSchemaChangeBackfillCheckpoint(t *testing.T) {
 			}
 			return errors.New("waiting for checkpoint")
 		})
+		t.Logf("initial checkpoint: %s", initialCheckpoint.Slice())
 
 		require.NoError(t, jobFeed.Pause())
 
@@ -2375,7 +2377,7 @@ func TestChangefeedSchemaChangeBackfillCheckpoint(t *testing.T) {
 			// Filter non-backfill-related spans
 			if !r.Timestamp.Equal(backfillTimestamp.Next()) {
 				// Only allow spans prior to a valid backfillTimestamp to avoid moving past the backfill
-				return !(backfillTimestamp.IsEmpty() || r.Timestamp.LessEq(backfillTimestamp.Next())), nil
+				return backfillTimestamp.IsSet() && !r.Timestamp.LessEq(backfillTimestamp.Next()), nil
 			}
 
 			require.Falsef(t, initialCheckpoint.Encloses(r.Span), "second backfill should not resolve checkpointed span")
@@ -2401,6 +2403,7 @@ func TestChangefeedSchemaChangeBackfillCheckpoint(t *testing.T) {
 			}
 			return errors.New("waiting for second checkpoint")
 		})
+		t.Logf("second checkpoint: %s", secondCheckpoint.Slice())
 
 		require.NoError(t, jobFeed.Pause())
 		for _, span := range initialCheckpoint.Slice() {
@@ -2420,10 +2423,11 @@ func TestChangefeedSchemaChangeBackfillCheckpoint(t *testing.T) {
 		// checkpoint should eventually be gone once backfill completes.
 		testutils.SucceedsSoon(t, func() error {
 			progress := loadProgress()
-			if p := progress.GetChangefeed(); p == nil || p.Checkpoint == nil || len(p.Checkpoint.Spans) == 0 {
-				return nil
+			if p := progress.GetChangefeed(); p != nil && p.Checkpoint != nil && len(p.Checkpoint.Spans) > 0 {
+				t.Logf("non-empty checkpoint: %s", progress.GetChangefeed().Checkpoint.Spans)
+				return errors.New("checkpoint still non-empty")
 			}
-			return errors.New("checkpoint still non-empty")
+			return nil
 		})
 
 		// Pause job to avoid race on the resolved array

